### PR TITLE
Add plan authz and set credentials

### PR DESCRIPTION
**Pull request by makyo**
Additionally, pass error information into the callback's error argument, rather than in data.

### CODE REVIEW

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js [hunk 1]
```diff
@@ -33,6 +33,34 @@ var module = module;
 
   plans.prototype = {
     /**
+      Get a plan authorization to deploy a charm.
+
+      @public authorizePlan
+      @params env-uuid {String} UUID of the environment the charm is being deployed on
```
Comment (frankban) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js]: Please lets call this modelUUID, or just uuid, and document the fact the it's the uuid of a model.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js [hunk 2]
```diff
@@ -33,6 +33,34 @@ var module = module;
...
+      @params charm-url {String} URL of the charm being deployed
```
Comment (frankban) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js]: s/charm-url/charmUrl/

Comment (frankban) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js]: And please complete the sentence, ere and below.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js [hunk 3]
```diff
@@ -33,6 +33,34 @@ var module = module;
...
+      @params service-name {String} name of the service
```
Comment (frankban) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js]: `@params applicationName {String} The name of the application.`

Comment (frankban) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js]: And please replace service with application anywhere ;-)

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js [hunk 4]
```diff
@@ -33,6 +33,34 @@ var module = module;
...
+      @params plan-url {String} url of the plan being deployed
```
Comment (frankban) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/plans.js]: s/plan-url/planUrl/ same for all dash separated fields below.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5253/
Test FAILed.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/test-plans.js [hunk 5]
```diff
@@ -204,4 +204,38 @@ describe('jujulib plans service', function() {
...
+      'charmUrl',
+      'serviceName',
+      'planUrl',
+      'budget',
+      'limit',
+      function(error, authz) {
+        assert.equal(error, null);
+        assert.equal(authz['look ma'], 'I\'m a macaroon');
+        assert.equal(authz.params,
+          '{"env-uuid":"envUUID","charm-url":"charmUrl",' +
```
Comment (frankban) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/test-plans.js]: Nice!

File: jujugui/static/gui/src/app/store/env/go.js [hunk 6]
```diff
@@ -3217,6 +3217,33 @@ YUI.add('juju-env-go', function(Y) {
         Request: request,
         Params: {Tag: userTag}
       }, handleListModels);
+    },
+
+    /**
+      setMetricCredentials sets credentials on an application.
```
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: This is already implemented :-/
Sorry, let's just pick the best impl here.

**Comment by frankban** [comment]
👍 with a couple of minor (mostly naming) changes.
Thanks a lot, and I am really sorry for the re-work being done as part of this branch.
Fix it then ship it!

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/index.js [hunk 7]
```diff
@@ -47,6 +47,11 @@ var module = module;
       if (parse !== false) {
         try {
           data = JSON.parse(data);
+          if (data.error) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/index.js]:
```JavaScript
error = data.error || data.Error;
```

File: jujugui/static/gui/src/app/store/env/go.js (hunk 6 above)
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: I like the existing one :)

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/index.js (hunk 7 above)
Comment (makyo) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/index.js]: Oh, cheers.  I was worried this would evaluate as boolean.

**Comment by makyo** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5255/
Test PASSed.

**Pull request merged**